### PR TITLE
Added completion for make

### LINKED ISSUE #128: Completion script for Makefiles?
Is it possible to make a completion script that will scan for Makefile targets and provide autocompletion for them? I've tried reading the Clink docs and some of the scripts in this repo, but I don't know enough Lua to make heads or tails of it.

If somebody who knows how clink/lua work can tell me if the idea is even reasonable, I'd really appreciate it.

### PR DESCRIPTION
**Pull request by Suraj-Yadav**
Overview:
1. Run entered make command with `-pqr` to generate database output. [Explaination](https://explainshell.com/explain?cmd=make+-pqr)
2. Parse output to find targets

Should fix #128

### CODE REVIEW

**Comment by chrisant996** [comment]
Interesting! Thanks for sharing this. I'll review it in about a week, once I'm available again.

- Do you know if `-pqrR` are available in non-gnu versions of `make`?
- Do you know if they're available in `nmake` from Visual Studio?

**Comment by Suraj-Yadav** [comment]
* I am not 100% sure about non-gnu versions of `make`. As per manpage [here](https://www.unix.com/man-page/posix/1p/make), only `pqr` are supported. But for `p` it says `The output format is unspecified`, so it is possible that target selection logic would not work or give targets which are not there.
* I tested this on `nmake` and again only `pqr` is available. Completion works but I can see an extra value `.SUFFIXES`.

Note `R` flag is not essential for this. It filters inbuilt variables, which we will anyway not capture.

**Comment by chrisant996** [comment]
> But for `p` it says `The output format is unspecified`, so it is possible that target selection logic would not work or give targets which are not there.

I don't think it's working ok with nmake.  For example, it lists `.SUFFIXES` as a target (a parsing problem) and it's listing every single intermediate file from implicit rules.  So e.g. `debug\foo.obj` and `debug\bar.obj` and so on.

Or is that intentional?

**Comment by Suraj-Yadav** [comment]
That is not intentional.

Though this is not a problem which is particular to `nmake`. I generated a Makefile for `make` using cmake, it is also listing targets like `debug\foo.obj` and `debug\bar.obj`.

I did a comparison of completion in `zsh` and `bash`.
**zsh** [Impl](https://github.com/zsh-users/zsh/blob/master/Completion/Unix/Command/_make)
image
**bash** [Impl](https://github.com/scop/bash-completion/blob/master/completions/make)
image
**My Implementation**
image

* Bash does filters all intermediate files. But this introduces a problem. If i have a manual target to generate a file, bash doesn't show that.
* zsh shows intermediate files. Though as a bonus it does allow completion of variables inside Makefile.
* My implementation incorrectly shows patterns (like `%`) as targets. 

I will try to remove patterns from my implementation this weekend.

What should be our target though? Like should we aim for `zsh` or `bash`.

**Comment by chrisant996** [comment]
@Suraj-Yadav On looking closer, I noticed that this uses `line_state:getline()` as the base command to run.

That's overly simplistic and highly dangerous -- there is no parsing or escaping or validation of the input -- it is used blindly, which makes it vulnerable to "injection attack".  For example:  An input line can contain multiple commands, and the `line_state:getline()` is the _full_ input line.  If the user types `del * & make `<kbd>Alt-=</kbd> then the script will run `del * & make -pqrR 2>nul` to get the targets.  Observe the extra `del * &` that does not belong as part of the command to run.  Also, any flags and redirection symbols the user types can interfere and cause the script to malfunction.

I assume the intent was probably to support input like `make -f mymakefile`?  The script cannot just pass in the user's typed input line.  The script must precisely parse the input line.  See [Responding to Arguments in Argmatchers](https://chrisant996.github.io/clink/clink.html#responsive-argmatchers) for background info on one way to handle that.

**Comment by chrisant996** [comment]
> What should be our target though? Like should we aim for `zsh` or `bash`.

Personally, I would probably intuitively expect to see "meta" targets -- things like "all" and "clean" and "exe", but not things like "debug\foo.exe" or "debug\bar.obj" or "src/file.cpp.i".

I'm not sure how the zsh implementation is coming up with variable names like `CMAKE_COMMAND`, and I don't know whether that's useful or accurate.

The output from the bash implementation looks roughly like what I would have intuitively expected to see.

It's probably a good idea to examine the bash implementation.  Maybe it spawns something like `make -pqrR` and filters away undesirable results, or maybe it parses makefiles directly -- I don't know.

**Comment by chrisant996** [comment]
I added [`completions\nmake.lua`](https://github.com/vladimir-kotikov/clink-completions/blob/master/completions/nmake.lua), based on the `-pqr` technique you've used here.

Can you use that as a template for making a `completions\make.lua` script?  I added a bunch of comments in the script to hopefully help clarify what it's doing and why, and I'm happy to answer any further questions.

**Comment by Suraj-Yadav** [comment]
@chrisant996 I have incorporated the requested changes.

There was one case I encountered ([here](https://github.com/vladimir-kotikov/clink-completions/pull/172/files#diff-61917990445e3fcaddd3d4b0fbb961bbf699438c13f0d5d56dd86edf1a35cc08R124)), which I am not sure is a bug or expected behavior. I worked around it, but please inform me if i missed something.

I have two questions
1. make has a `-C` flag. `-C DIRECTORY, --directory=DIRECTORY Change to DIRECTORY before doing anything.` 
   `help_parser` recognizes this flag but doesn't assigns it `clink.dirmatches`, like it assigns `clink.filematches` for `-f`. 
   Should `DIRECTORY` be added to `_dir_keywords` [here](https://github.com/vladimir-kotikov/clink-completions/blame/master/modules/help_parser.lua#L97)?
2. How do I test changes to completion easily?
   Currently I save the file, then open a new `cmd` and then verify if completions are working or not.

**Comment by chrisant996** [comment]
I'll review the changes this afternoon.  In the meantime, some answers to the questions:

> There was one case I encountered ([here](https://github.com/vladimir-kotikov/clink-completions/pull/172/files#diff-61917990445e3fcaddd3d4b0fbb961bbf699438c13f0d5d56dd86edf1a35cc08R124)), which I am not sure is a bug or expected behavior. I worked around it, but please inform me if i missed something.

Oh...  That's an unforeseen side effect of the argmatcher feature that supports flags like `--blah:param` (and `--blah:c:\foo\bar` works fine, btw).

Clink's line parsing happens before -- and completely independently from -- any argmatchers are evaluated, so it has no idea whether `-fc:\foo\bar` is a flag `-fc:` followed by `\foo\bar` or a flag `-f` followed by `c:\foo\bar`.  When I was hooking up support for flags that end with `:` or `=` I overlooked that there are some flags that accept a parameter directly appended to the flag, without a `:` or `=`.

I'll explore options for whether/how this could be handled better.  But for now, argmatchers must compensate on their own.

> 1. make has a `-C` flag. `-C DIRECTORY, --directory=DIRECTORY Change to DIRECTORY before doing anything.`
>    `help_parser` recognizes this flag but doesn't assigns it `clink.dirmatches`, like it assigns `clink.filematches` for `-f`.
>    Should `DIRECTORY` be added to `_dir_keywords` [here](https://github.com/vladimir-kotikov/clink-completions/blame/master/modules/help_parser.lua#L97)?

Yes.  I'll take care of that, since I'll need to push the updated help_parser.lua file to a couple of other places as well.

> 2. How do I test changes to completion easily?
>    Currently I save the file, then open a new `cmd` and then verify if completions are working or not.

Do you know about <kbd>Ctrl</kbd>-<kbd>x</kbd>,<kbd>Ctrl</kbd>-<kbd>r</kbd>?
By default that key sequence is bound to `clink-reload`, which reloads Clink (including .inputrc file and Lua scripts).

So there's no need to open a new `cmd` window.  Save the Lua script(s), and then in the Clink cmd window press <kbd>Ctrl</kbd>-<kbd>x</kbd>,<kbd>Ctrl</kbd>-<kbd>r</kbd> to reload Clink, then test whatever is being tested.  (I do that _a lot_ as you can imagine. 😉🤣)

File: completions/make.lua [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    ['.SILENT'] = true,
+    ['.EXPORT_ALL_VARIABLES'] = true,
+    ['.NOTPARALLEL'] = true,
+    ['.ONESHELL'] = true,
+    ['.POSIX'] = true,
+    ['.NOEXPORT'] = true,
+    ['.MAKE'] = true,
+}
+
+local function extract_target(line, last_line, targets, allow_pathlike)
```
Comment (chrisant996) [inline completions/make.lua:28]: Can you add (back) a comment describing the `extract_target` function?

I would like for diffing nmake.lua vs make.lua to show necessary functional changes.  There are a lot of whitespace changes, removed comments, and some renamed variables.  After this PR is merged, I'll resolve the whitespace changes and renamed variables so they're consistent between the two files.

File: completions/make.lua [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    -- Extract possible target.
+    local p = (
+        line:match('^([^%s]+):$') or -- When target has no deps.
+        line:match('^([^%s]+): '))   -- When target has deps.
+    if not p then
+        return
+    end
+
+    -- Ignore targets with pattern
```
Comment (chrisant996) [inline completions/make.lua:47]: Is "pattern" a typo?  Was it meant to be "percent"?

File: completions/make.lua [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    local a_alias = (a.type == 'alias')
+    local b_alias = (b.type == 'alias')
+    if a_alias ~= b_alias then
+        return a_alias
+    else
+        return string.comparematches(a.match, b.match)
+    end
+end
+
+local function get_targets(word, word_index, line_state, builder, user_data) -- luacheck: no unused
```
Comment (chrisant996) [inline completions/make.lua:84]: The reason I used `_word` and `_word_index` in the nmake.lua script is because I observed that in your original make.lua script unused param names were prefixed with `_`.  I was trying to accept that convention, though I don't prefer it.

Should I abandon that convention in both the make.lua and nmake.lua scripts?

File: completions/make.lua [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    local file = io.popen('2>nul ' .. make_cmd)
+    if not file then
+        return
+    end
+
+    local targets = {}
+    local last_line = ''
+
+    local allow_pathlike = os.getenv('INCLUDE_PATHLIKE_MAKEFILE_TARGETS') and true
```
Comment (chrisant996) [inline completions/make.lua:98]: I can understand why you might prefer a longer name than `pathlike`.

I have two observations about the changed name:
1. The name change makes the scripts diverge unnecessarily in diffs.  I would like comparing the two scripts to show functional changes (without stylistic whitespace changes and name changes and comment removals), and I would like for both scripts to use the same name.
2. The name introduces a new word "allow", but the env var uses "include".  Can the name be `include_pathlike`?

I'm willing to update nmake.lua accordingly after this PR is merged.

File: completions/make.lua [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        if builder.setnosort then
+            builder:setnosort()
+        end
+    end
+
+
+    return targets
+end
+
+local function onarg_flags(arg_index, word, word_index, line_state, user_data)
```
Comment (chrisant996) [inline completions/make.lua:120]: This function has been moved.  Can you keep the code in the same relative order as in the nmake.lua file, so that comparing the files has fewer unnecessary diff chunks?

File: completions/make.lua [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+-- This is empty as all flags for make are parsed by help_parser
+local flags_table = {}
+-- Add onarg function to detect when the user overrides the default makefile.
+flags_table.onarg = onarg_flags
+
+
+local make_parser = clink.argmatcher("make")
+make_parser:_addexflags(flags_table)
+make_parser:addarg({ get_targets })
+make_parser:nofiles()
```
Comment (chrisant996) [inline completions/make.lua:143]: The repeated use of `make_parser` in lines 141-143 is unnecessary.  Each of the functions returns the argmatcher, which is why they can be chained like `clink.argmatcher():addarg():nofiles():loop():addflags()` etc.

**Review by chrisant996 — CHANGES_REQUESTED** [review]
Looks good.  I added feedback about a few quick things.

By the way, it seems the file now uses a different spacing style than before.  I'd like to be consistent at least between the nmake.lua and make.lua scripts.
- I'm ok with either `a..b` or `a .. b` but would like to maintain consistency, and would like to avoid changing them from one commit to another.
- I'm ok with either `{a=x, b=y}` or `{ a=x, b=y }` or `{ a = x, b = y }` but would like to maintain consistency etc.

Which styles will you be using in this file?

File: completions/make.lua (hunk 2 above)
Comment (Suraj-Yadav) [reply completions/make.lua:47]: It is pattern.
https://www.gnu.org/software/make/manual/html_node/Pattern-Intro.html

File: completions/make.lua (hunk 3 above)
Comment (Suraj-Yadav) [reply completions/make.lua:84]: This is my first time working with lua, so I was not sure what the general convention is and don't have any preference.
I think we can abandon _ convention, the existing files don't use that (which I should have consulted in the first place😅)

**Review by chrisant996 — APPROVED** [review]
👍👍
